Links.md

```diff
@@ -0,0 +1,75 @@
+# Link encoding in IPLD
+
+*Draft* notes and recommendations for link encoding and decoding across IPLD
+codec implementations.
+
+## Terms
+
+A logical separation exists in any given IPLD codec between the **format** and the **serializer/deserializer**.
+
+```
++--------------------+             +---------------------+
+|                    |             |                     |
+|     Serializer     |             |    Deserializer     |
+|                    |             |                     |
++---------+----------+             +----------+----------+
+          |                                   ^
+          |         Sent to another peer      |
+          v                                   |
++---------+----------+             +----------+----------+
+|                    |             |                     |
+|      Format        +------------->       Format        |
+|                    |             |                     |
++--------------------+             +---------------------+
+```
+
+A codec may represent object types and tree structures any way it wishes.
+This includes existing representations (JSON, BSON, CBOR, Protobuf, msgpack,
+etc) or even new custom serializations. We will refer to this as the
+**representation**.
+
+Therefor, a **format** is the standardized representation of IPLD Links and Paths in a given **representation**.
+
+It is worth noting that **serializers** and **deserializers** differ by programming language while the **format** does not and MUST remain consistent across all codec implementations.
+
+# Canonical Link Representation
+
+Codec **serializers** MUST reserve the following canonical
+representation of link encoding. The canonical representation is an object with a single key of `"/"` and a base encoded string of the link's CID.
```

Comment:
We originally said that no objects can have slashes in keys (keys must be valid path components) but backed off when we realized that wasn't going to work. At this point, I'm not sure if we *can* introduce a restriction like this. CBOR objects definitely can have a single "/" keys.

We really do need to sit down and think through what can and can't go into an IPLD object because I think we're getting closer and closer to "everything goes". That might be fine but we need to address it explicitly...

Reply:
Also note, this was *never* intended to be the canonical representation. It was a hack to get JSON working.

Reply:
I was somewhat aware of the history. But I think that we do need *some* form of canonical representation that can be represented in pure JSON in order to open a path for people to encode objects from one format to another.

However, I don't think, and am actively trying to change in `dag-cbor`, the default use of the canonical representation in the deserializer. It's a horrible pain to work with and, while I want to reserve it for interop, I don't want it to be in common use but instead buried in the implementations.

Reply:
> We originally said that no objects can have slashes in keys (keys must be valid path components) but backed off when we realized that wasn't going to work

It is worth nothing that CID encoded in base64 will have slashes in them.  I am wondering it was really a good idea to allow that encoding, it will mess up paths like `/ipfs/<base 64 cid>/file.txt`.